Add instagram login to master (with google and facebook)

ferdefiore opened the pull request:
The three logins are working well. About what i did, instagram do not have a framework so i had to use a service through a web view so i added an mvp for instagram and some utils classes that i use eg for take the users information, etc.

ferdefiore committed:
ig, google and facebook login working

gregogerardi commented on app/src/main/java/com/pps/globant/fittracker/MainActivity.java:
```diff
@@ -48,4 +60,15 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
             callbackManager.onActivityResult(requestCode, resultCode, data);
     }
 
+    @OnClick(R.id.btn_insta_login)
```
this should be into the view

SartiMau commented on app/src/main/java/com/pps/globant/fittracker/MainActivity.java:
```diff
@@ -13,25 +15,35 @@
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
         ButterKnife.bind(this);
+        //presenter = new LoginPresenter(new LoginModel(), new LoginView(this));
```
remove comment please

ferdefiore replied:
done!

SartiMau commented on app/src/main/java/com/pps/globant/fittracker/mvp/presenter/InstagramLoginPresenter.java:
```diff
@@ -0,0 +1,86 @@
+    public void handleUrl(String url) {
+        String code;
+        if (url.contains(CODE)) {
+            String temp[] = url.split("=");
```
put this line before the if and remove it from the else statement.. and use constants please..

ferdefiore replied:
done!

SartiMau commented on app/src/main/java/com/pps/globant/fittracker/mvp/view/LoginView.java:
```diff
@@ -107,4 +111,37 @@ public void popUp(@StringRes int error) {
+    }
+
+    public void showLoginToast(String name){
+        Toast.makeText(this.getContext(), getContext().getResources().getString(R.string.toast_instagram_login) + name, Toast.LENGTH_SHORT).show();
```
It is necessary use "getContext().getResources().getString(R.string.toast_instagram_login)" isn't work with "R.string.toast_instagram_login"??

ferdefiore replied:
Yes! If i didnt use "getContext().getResources().getString(R.string.toast_instagram_login)" when the toast was shown it contains a number. This was the only case that i had to use it in that way, so i think its because of "+ name "

SartiMau replied:
ok np

nacho-globant replied:
You need use String.format instead the +

ferdefiore replied:
done!

SartiMau commented on app/src/main/java/com/pps/globant/fittracker/utils/StreamToString.java:
```diff
@@ -0,0 +1,31 @@
+public class StreamToString {
+
+    public static String streamToString(InputStream is) throws IOException {
+        String str = "";
```
Use constants please

ferdefiore replied:
done!

SartiMau commented on app/src/main/res/values/styles.xml:
```diff
@@ -25,4 +25,13 @@
 
+    <!--Instagram styles -->
+    <style name="instagramButton" parent= "AppTheme">
+        <item name="android:layout_width">277dp</item>
```
Use dimen please

ferdefiore replied:
done!

ferdefiore committed:
Apply what Mauro suggested to change:
-remove a comment
-use constants
-adjust an if sentence

gregogerardi commented on app/src/main/java/com/pps/globant/fittracker/utils/MyAsyncTask.java:
```diff
@@ -0,0 +1,83 @@
+import static com.pps.globant.fittracker.utils.CONSTANTS.SP_NAME;
+import static com.pps.globant.fittracker.utils.CONSTANTS.SP_TOKEN;
+
+public class MyAsyncTask extends AsyncTask<URL, Integer, Long> {
```
parameter to instance the AsyncTask<URL, Integer, Long are wrong. you don't need a Long as result, you didn't ever use it. same to the Integer second paramer for onProgress () method. instead of saving a private variable for the full name return it in doInBackground and accesing it in onPostExecute.

ferdefiore replied:
ok! thanks

gregogerardi commented on app/src/main/java/com/pps/globant/fittracker/mvp/presenter/InstagramLoginPresenter.java:
```diff
@@ -0,0 +1,85 @@
+    @Subscribe
+    public void onRetIgInformation(InstagramLoginModel.RetIgInformation event) {
+        if (dialog != null && dialog.isShowing()) {
+            dialog.dismiss();
```
dialog into the view

ferdefiore replied:
done

nacho-globant commented on app/src/main/java/com/pps/globant/fittracker/MainActivity.java:
```diff
@@ -13,25 +15,34 @@
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
```
This is not a error

ferdefiore replied:
That part of code is the same as the first commit, and it doesn't work without "R." maybe your IDEA is hiding it

gregogerardi replied:
ok, thanks for the clarification

nacho-globant commented on app/src/main/java/com/pps/globant/fittracker/MainActivity.java:
```diff
@@ -13,25 +15,34 @@
     private CallbackManager callbackManager;
     private LoginPresenter presenter;
     private static final int RC_GET_TOKEN = 9002;
+    SharedPreferences spUser;
```
Use this only as local variable

ferdefiore replied:
done

nacho-globant commented on app/src/main/java/com/pps/globant/fittracker/mvp/model/InstagramLoginModel.java:
```diff
@@ -0,0 +1,75 @@
+public class InstagramLoginModel {
+
+    private final static String AUTH_URL_FULL = AUTHURL + "client_id=" + CLIENT_ID + "&redirect_uri=" + REDIRECT_URI + "&response_type=code&display=touch";
+    private static String TOKEN_URL_FULL = TOKENURL + "?client_id=" + CLIENT_ID + "&client_secret=" + CLIENT_SECRET + "&redirect_uri=" + REDIRECT_URI + "&grant_type=authorization_code";
```
great comment 👍

ferdefiore replied:
done

nacho-globant commented on app/src/main/java/com/pps/globant/fittracker/mvp/model/InstagramLoginModel.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    public static String getTokenUrlFull() {
+        return TOKEN_URL_FULL;
```
`TOKEN_URL_FULL`is a constant you dont need specify a getter to use it, only mark as public

ferdefiore replied:
done

nacho-globant commented on app/src/main/java/com/pps/globant/fittracker/mvp/model/InstagramLoginModel.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    public String getAuthUrlFull() {
+        return AUTH_URL_FULL;
```
`AUTH_URL_FULL`  is a constant you dont need specify a getter to use it, only mark as public

ferdefiore replied:
done

nacho-globant commented on app/src/main/java/com/pps/globant/fittracker/mvp/model/InstagramLoginModel.java:
```diff
@@ -0,0 +1,75 @@
+        new MyAsyncTask(code,bus,spUser).execute();
+    }
+
+    public void isLoggedIn(MainActivity activity) {
```
all the logic should be into the presenter

ferdefiore replied:
moved!

nacho-globant commented on app/src/main/java/com/pps/globant/fittracker/mvp/presenter/InstagramLoginPresenter.java:
```diff
@@ -0,0 +1,85 @@
+    private final InstagramLoginModel model;
+    private final IntagramLoginView view;
+    public Dialog dialog;
+    public MainActivity activity;
```
remove this pls, you should access by view.getActivity()

ferdefiore replied:
done

nacho-globant commented on app/src/main/java/com/pps/globant/fittracker/mvp/presenter/InstagramLoginPresenter.java:
```diff
@@ -0,0 +1,85 @@
+        this.bus = bus;
+        BusProvider.register(this);
+        model = new InstagramLoginModel(bus);
+        view = new IntagramLoginView();
```
The view and model should be pass as parameter

ferdefiore replied:
done

nacho-globant commented on app/src/main/java/com/pps/globant/fittracker/mvp/presenter/InstagramLoginPresenter.java:
```diff
@@ -0,0 +1,85 @@
+        webView.setWebViewClient(new MyWVClient(dialog, bus));
+        webView.getSettings().setJavaScriptEnabled(true);
+        webView.loadUrl(model.getAuthUrlFull());
+        dialog.setContentView(webView);
```
You could create a new method into the view to do that. I think

ferdefiore replied:
done

nacho-globant commented on app/src/main/java/com/pps/globant/fittracker/mvp/presenter/InstagramLoginPresenter.java:
```diff
@@ -0,0 +1,85 @@
+    public void igButtonLoginClick(MainActivity activity) {
+        this.activity = activity;
+        dialog = new Dialog(activity);
+        WebView webView = new WebView(activity);
```
Why do you need a WebView?

ferdefiore replied:
i will commit in this pull request again with the webView, having fixed all the others reviews, to dont lose more time with the login and can start with the API but however i will study the implementation with an intent opening a browser like you recommend for another PR

nacho-globant commented on app/src/main/java/com/pps/globant/fittracker/mvp/presenter/InstagramLoginPresenter.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+
+    public void handleUrl(String url) {
```
I dont understand the idea for that method

ferdefiore replied:
I removed the method and put the code where the method was called. It was done in that way by an old implementation where was more complex, thanks

nacho-globant commented on app/src/main/java/com/pps/globant/fittracker/mvp/presenter/InstagramLoginPresenter.java:
```diff
@@ -0,0 +1,85 @@
+    @Subscribe
+    public void onPostForHandleUrl(MyWVClient.PostForHandleUrl event) {
+        String url = event.url;
+        handleUrl(url);
```
simplify the code and change by ;)        
        `handleUrl(event.url);`

ferdefiore replied:
ok!

ferdefiore commented on app/src/main/java/com/pps/globant/fittracker/mvp/presenter/LoginPresenter.java:
```diff
@@ -25,17 +29,22 @@
     private final LoginView view;
     private final Activity activity;
+    //instagram presenter
+    private final InstagramLoginPresenter IgPresenter = new InstagramLoginPresenter(BusProvider.getInstance());
```
done

ferdefiore commented on app/src/main/java/com/pps/globant/fittracker/mvp/model/InstagramLoginModel.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    public void isLoggedIn(MainActivity activity) {
+        spUser= activity.getSharedPreferences(SP,MODE_PRIVATE);
```
done

ferdefiore commented on app/src/main/res/values/dimens.xml:
```diff
@@ -11,5 +11,6 @@
 
-
+    <dimen name="instagram_name_text_size">20sp</dimen>
+    <dimen name="instagram_button_width">277</dimen>
```
ok

ferdefiore commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -58,11 +66,28 @@
+        <Button
+            android:id="@+id/btn_insta_login"
+            style="@style/instagramButton"
+            android:layout_height="wrap_content"
```
ok

ferdefiore commented on app/src/main/java/com/pps/globant/fittracker/mvp/presenter/InstagramLoginPresenter.java:
```diff
@@ -0,0 +1,85 @@
+        this.bus = bus;
+        BusProvider.register(this);
+        model = new InstagramLoginModel(bus);
+        view = new IntagramLoginView();
```
Is not used, it is here since i created the mvp for IG. I will remove it.

priettt replied:
are we correcting grammar?

gregogerardi replied:
> are we correcting grammar?

the buttton is a new component, more beauty and efficient than the button. you should research about it

ferdefiore commented on app/src/main/java/com/pps/globant/fittracker/mvp/model/InstagramLoginModel.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    @Subscribe
+    public void onPostReadyInformation(MyAsyncTask.PostReadyInformation event){
```
Why? This is used in the InstagramLoginModel togheter with MyAsyncTask that gets the user information. If what I understand is not wrong, i cant pass over the model to the presenter directly with the AsyncTask.
The Instagram login model has another one for return information to the presenter.

gregogerardi replied:
all logic should go into the pressenter, subscribe the presenter to the event and then if some accions are needed into the model call them from the pressenter. then the model could send another evento to the presenter. refer to a previous correccion from nacho in my fb login pr where he told me the same https://github.com/priettt/FitTracker/pull/4#discussion_r221687138 or also https://github.com/priettt/FitTracker/pull/4#discussion_r221687237

ferdefiore replied:
I understand what you say. I asked to @capraber when i did this, if i could pass step over the model directly from MyAsyncTask to the presenter, because it is used for this, it is like an extra step.. but he said that i couldnt, that it should be MyAsyncTask --> InstagramLoginModel --> InstagramLoginPresenter, so i did it in that way, maybe my question was not good, so he didnt understood

ferdefiore commented on app/src/main/java/com/pps/globant/fittracker/utils/MyWVClient.java:
```diff
@@ -0,0 +1,61 @@
+        super.onPageStarted(view, url, favicon);
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
```
changed

ferdefiore commented on app/src/main/java/com/pps/globant/fittracker/utils/MyWVClient.java:
```diff
@@ -0,0 +1,61 @@
+
+    @Override
+    public void onPageStarted(WebView view, String url, Bitmap favicon) {
+        super.onPageStarted(view, url, favicon);
```
ok

ferdefiore committed:
Some corrections before use the IgLoginView

ferdefiore committed:
Some changes before use the IgLoginView

ferdefiore commented on app/src/main/java/com/pps/globant/fittracker/utils/MyWVClient.java:
```diff
@@ -0,0 +1,61 @@
+    @Override
+    public void onPageFinished(WebView view, String url) {
+        super.onPageFinished(view, url);
+        dialog.show();
```
moved to the IgLoginView

ferdefiore commented on app/src/main/java/com/pps/globant/fittracker/mvp/presenter/InstagramLoginPresenter.java:
```diff
@@ -0,0 +1,85 @@
+        String temp[] = url.split(EQUAL_SIGN);
+        if (url.contains(CODE)) {
+            code = temp[1];
+            model.getIGInformation(code, activity);
```
cause an old implementation, now its clean, thanks

ferdefiore committed:
-All reviews fixed
-Version with WebView yet, nacho said that i should use an intent with a browser, i will do it later in another PR
-----
-Moved some code between classes, related with mvp cause were in wrongs places
-Added the use of String.format

ferdefiore committed:
-Version with WebView yet, nacho said that i should use an intent with a browser, i will do it later in another PR
-----
-Moved some code between classes, related with mvp cause were in wrongs places
-Added the use of String.format

+

Check view.getActivity() != null

capraber reviewed (approved):
+1
